trl/experimental/gold/gold_trainer.py

```diff
@@ -0,0 +1,1984 @@
+
+            # Also mask pad tokens in teacher labels for consistency
+            if (
+                hasattr(self, "teacher_tokenizer")
+                and hasattr(self.teacher_tokenizer, "pad_token_id")
+                and self.teacher_tokenizer.pad_token_id is not None
+            ):
+                teacher_labels[teacher_labels == self.teacher_tokenizer.pad_token_id] = -100
+
+            loss = self.uld_loss_fn(
+                student_logits=outputs_student.logits,
+                teacher_logits=outputs_teacher.logits,
+                student_labels=student_labels,
+                teacher_labels=teacher_labels_for_loss,
+                student_input_ids=student_input_ids,
+                teacher_input_ids=teacher_input_ids_for_loss,
+            )
+
+            # If ULD hybrid mode produced per-step matched/unmatched components, accumulate them for logging.
+            # Use gradient_accumulation_steps to mirror Trainer's windowing behavior.
+            if hasattr(self.uld_loss_fn, "last_matched_loss") and hasattr(self.uld_loss_fn, "last_unmatched_loss"):
+                try:
+                    ga = max(1, int(self.args.gradient_accumulation_steps))
+                except Exception:
+                    ga = 1
+                step_eq = 1.0 / ga
+                # read scalar values for logging
+                matched_val = (
+                    self.uld_loss_fn.last_matched_loss.item()
+                    if self.uld_loss_fn.last_matched_loss is not None
+                    else 0.0
+                )
+                unmatched_val = (
+                    self.uld_loss_fn.last_unmatched_loss.item()
+                    if self.uld_loss_fn.last_unmatched_loss is not None
+                    else 0.0
+                )
+
+                self._matched_sum += matched_val
+                self._unmatched_sum += unmatched_val
+                self._matched_step_eq += step_eq
+                self._unmatched_step_eq += step_eq
+
+        empty_cache()
+
+        return (loss, outputs_student) if return_outputs else loss
+
+    def generate_on_policy_outputs(self, model, inputs, generation_config, pad_token_id=None):
+        # Generate output with respect to the prompt only
+        if self.use_transformers_paged:
+            previous_attn = self.model.config._attn_implementation
+            if is_flash_attn_2_available():
+                model.config._attn_implementation = "paged_attention"
+            else:
+                model.config._attn_implementation = "sdpa_paged"
+            generated_outputs = model.generate_batch(
+                inputs["prompts"].tolist(),
+                generation_config=generation_config,
+            )
+            model.config._attn_implementation = previous_attn
+
+            completion_ids = [output.generated_tokens for output in generated_outputs.values()]
+            generated_tokens = torch.stack([torch.tensor(ids, device=model.device) for ids in completion_ids])
```

Comment:
I'm quite surprised that this doesn't fail, because it's likely that most completions don't share the same length

Reply:
so here we are in the generate_batch part that returns a dict whose values are already right padded to the same length so it doesnt fail